core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -206,8 +206,13 @@ protected TransportRequestOptions transportOptions() {
     }
 
     protected boolean retryPrimaryException(final Throwable e) {
-        return e.getClass() == ReplicationOperation.RetryOnPrimaryException.class
-            || TransportActions.isShardNotAvailableException(e);
+        boolean retry = e.getClass() == ReplicationOperation.RetryOnPrimaryException.class
+                || TransportActions.isShardNotAvailableException(e);
+        if (retry) {
+            assert e instanceof ElasticsearchException
```

Comment:
This was to ensure all retryPrimaryExceptions are indeed instance of ElasticsearchException, I added this to make sure none of the retryPrimaryExceptions were silently shallowed at TransportShardBulkAction. I reverted this change. see https://github.com/elastic/elasticsearch/pull/19105#r73494534 for previous discussion on this.